### PR TITLE
Add extend/os/linux/system_config.rb

### PR DESCRIPTION
**Pull request by sjackman**
Report the Linux kernel, OS, and gcc versions.

- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/master/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/homebrew/pull/49031).
- [ ] Have you successfully run `brew tests` with your changes locally?

### CODE REVIEW

File: Library/Homebrew/extend/os/linux/system_config.rb [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+    end
+
+    def host_gcc_version
+      gcc = Pathname.new "/usr/bin/gcc"
+      `#{gcc} --version 2>/dev/null`[/ (\d+\.\d+\.\d+)/, 1] if gcc.exist?
+    end
+
+    def formula_version(formula)
+      f = Formula[formula]
+      f.installed? ? f.version : "N/A"
```
Comment (MikeMcQuaid) [inline Library/Homebrew/extend/os/linux/system_config.rb]:
```
return "N/A" unless f.installed?
f.version
```

File: Library/Homebrew/extend/os/linux/system_config.rb [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    rescue FormulaUnavailableError
+      return "N/A"
+    end
+
+    def dump_verbose_config(f = $stdout)
+      dump_generic_verbose_config(f)
+      f.puts "Kernel: #{`uname -mors`.chomp}"
+      f.puts "OS: #{host_os_version}"
+      f.puts "OS gcc: #{host_gcc_version}"
```
Comment (MikeMcQuaid) [inline Library/Homebrew/extend/os/linux/system_config.rb]: Maybe `/usr/bin/gcc` would be more obvious?

File: Library/Homebrew/extend/os/linux/system_config.rb [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+      f = Formula[formula]
+      f.installed? ? f.version : "N/A"
+    rescue FormulaUnavailableError
+      return "N/A"
+    end
+
+    def dump_verbose_config(f = $stdout)
+      dump_generic_verbose_config(f)
+      f.puts "Kernel: #{`uname -mors`.chomp}"
+      f.puts "OS: #{host_os_version}"
```
Comment (MikeMcQuaid) [inline Library/Homebrew/extend/os/linux/system_config.rb]: `OS version`?

File: Library/Homebrew/extend/os/linux/system_config.rb [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+      f.puts "OS gcc: #{host_gcc_version}"
+      f.puts "Linuxbrew glibc: #{formula_version "glibc"}"
```
Comment (MikeMcQuaid) [inline Library/Homebrew/extend/os/linux/system_config.rb]: `glibc installed formula` or something?

Comment (MikeMcQuaid) [inline Library/Homebrew/extend/os/linux/system_config.rb]: Also I believe it's `Glibc`

File: Library/Homebrew/extend/os/linux/system_config.rb [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+      f.puts "Linuxbrew gcc: #{formula_version "gcc"}"
```
Comment (MikeMcQuaid) [inline Library/Homebrew/extend/os/linux/system_config.rb]: `GCC`?

File: Library/Homebrew/extend/os/linux/system_config.rb [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+      f.puts "Linuxbrew xorg: #{formula_version "linuxbrew/xorg/xorg"}"
```
Comment (MikeMcQuaid) [inline Library/Homebrew/extend/os/linux/system_config.rb]: `Xorg`?

File: Library/Homebrew/extend/os/linux/system_config.rb (hunk 3 above)
Comment (sjackman) [inline Library/Homebrew/extend/os/linux/system_config.rb]: I'm mimicking the output of `brew config` on macOS where it reads:
```
macOS: 10.11.6-x86_64
```
so I went for
```
OS: Ubuntu 14.04.5 LTS (trusty)
```
I feel the "version" is implied.

File: Library/Homebrew/extend/os/linux/system_config.rb (hunk 2 above)
Comment (sjackman) [inline Library/Homebrew/extend/os/linux/system_config.rb]: Agreed.

File: Library/Homebrew/extend/os/linux/system_config.rb (hunk 5 above)
Comment (sjackman) [inline Library/Homebrew/extend/os/linux/system_config.rb]: I prefer using the name of the formula `gcc` as opposed to the upstream name.

File: Library/Homebrew/extend/os/linux/system_config.rb (hunk 4 above)
Comment (sjackman) [inline Library/Homebrew/extend/os/linux/system_config.rb]: I prefer to use the name of the formula `glibc` as opposed to the upstream name.

File: Library/Homebrew/extend/os/linux/system_config.rb (hunk 5 above)
Comment (sjackman) [inline Library/Homebrew/extend/os/linux/system_config.rb]: How about `homebrew/core/gcc` ?

File: Library/Homebrew/extend/os/linux/system_config.rb (hunk 6 above)
Comment (sjackman) [inline Library/Homebrew/extend/os/linux/system_config.rb]: How about `How about `linuxbrew/xorg/xorg` ?

File: Library/Homebrew/extend/os/linux/system_config.rb (hunk 4 above)
Comment (MikeMcQuaid) [inline Library/Homebrew/extend/os/linux/system_config.rb]: How about just `glibc`?

File: Library/Homebrew/extend/os/linux/system_config.rb (hunk 6 above)
Comment (MikeMcQuaid) [inline Library/Homebrew/extend/os/linux/system_config.rb]: How about just `xorg`?

File: Library/Homebrew/extend/os/linux/system_config.rb (hunk 5 above)
Comment (MikeMcQuaid) [inline Library/Homebrew/extend/os/linux/system_config.rb]: How about just `gcc`?

**Comment by MikeMcQuaid** [comment]
Trying to think of a way of not using `Linuxbrew` in these names as Homebrew on Linux doesn't imply Linuxbrew (yet, at least) and these technically may use any tap's formula.

**Comment by sjackman** [comment]
Thanks for the suggestions, Mike. I've updated the PR.

**Comment by MikeMcQuaid** [comment]
CI failing; may need a more aggressive `rescue`: https://travis-ci.org/Homebrew/brew/jobs/211025276#L1036

**Comment by sjackman** [comment]
The error is
```
Error: Invalid formula: /home/travis/build/Homebrew/brew/Library/Taps/homebrew/homebrew-core/Formula/media-info.rb
uninitialized constant Formulary::FormulaNamespacea2b7870269c6d817b9989c0f5561f49f::GobjectIntrospection::MacOS
```
https://travis-ci.org/Homebrew/brew/jobs/211025276#L296
caused by for example
```ruby
depends_on "curl" if MacOS.version < :snow_leopard
```
because the constant MacOS is not defined when Homebrew is run on Linux.

The failing command is `brew config`. I'm guessing that `Formula["gcc"]` is causing `brew` to tap `Homebrew/core` and check the syntax validity of all those formula. If my understanding above is correct, options are…

- We could load the actual `MacOS` constant on Linux.
- We could load a stubbed out version of the `MacOS` constant that is just for Linux.
- We could somehow prevent `homebrew/core` from being loaded on Linux, by for example forcing the use of `NullLoader` for commands other than just `brew tests`.
- We could catch and ignore the `Invalid formula` exception in `brew config` alongside ignoring the `FormulaUnavailableError` exception.

Thoughts on which of these possible solutions is preferable?

**Comment by sjackman** [comment]
I'm overcomplicating the situation I think. I only need to use an equivalent of `Formulary.factory` that does not auto-tap Homebrew/core if it's not yet tapped.

**Comment by sjackman** [comment]
Fixed it (hopefully) with
```ruby
return "N/A" unless CoreTap.instance.installed?
```

**Comment by MikeMcQuaid** [comment]
Thanks again @sjackman!

**Comment by sjackman** [comment]
Thanks for merging, Mike!

**Pull request merged**